woocommerce/woocommerce-ios
# External products: edit SKU

Edit SKU for #2200

- [x] ⚠️ Make sure PR base is `develop` after https://github.com/woocommerce/woocommerce-ios/pull/2432 is merged ⚠️ 

## Changes

Originally, I created a new view controller for the Edit SKU screen as the SKU field design was a bit different from the inventory settings. Then I clarified with G, and we are reusing the SKU field UI for both Edit Inventory Settings (SKU plus other inventory fields) and Edit SKU screens. Since there is error handling when the SKU is already used by another product, I decided to reuse `ProductInventorySettingsViewController` for Edit SKU screen. I'm open to any other suggestions!

- In `ProductInventorySettingsViewController`, added an enum `FormType` that indicates the type of form this view controller is used for: inventory settings (SKU plus other fields), or SKU form (SKU field only). The differences between these 2 form types are:
  - Navigation bar title
  - In `viewDidAppear`, the keyboard is prompted for Edit SKU form
  - The sections for the table view: Edit SKU form only has the SKU section
- In `ProductFormViewModel`, added a function to update SKU + test case
- In `ProductFormViewController`, implemented SKU editing action

## Testing

- Go to the Products tab
- Tap on an external/affiliate Product
- If the product already has a SKU, tap on the SKU row; otherwise, tap on the "Add more details" CTA at the bottom, then tap on the SKU row --> the SKU screen should appear with keyboard up
- Enter a SKU that is used by another product --> an error should appear
- Enter a new valid SKU and tap "Done"
- Tap "Update" --> the product SKU should be updated remotely
- Go back to the Products tab
- Tap on a simple product
- Tap on the inventory row --> the inventory settings screen should appear without the keyboard up

## Example screenshots/screencast

![simulator](https://user-images.githubusercontent.com/1945542/84845253-c07ce580-b07e-11ea-8bca-17edf31061ef.gif)

existing SKU | empty SKU | duplicate SKU
-- | -- | --
![Simulator Screen Shot - iPhone 11 - 2020-06-17 at 09 16 10](https://user-images.githubusercontent.com/1945542/84845123-7a278680-b07e-11ea-8b59-95c65409773a.png) | ![Simulator Screen Shot - iPhone 11 - 2020-06-17 at 09 16 15](https://user-images.githubusercontent.com/1945542/84845158-86abdf00-b07e-11ea-9cf5-c79b387b26f4.png) | ![Simulator Screen Shot - iPhone 11 - 2020-06-17 at 09 16 23](https://user-images.githubusercontent.com/1945542/84845189-99261880-b07e-11ea-8059-05eaa765f2a7.png)


Update release notes:
